server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java

```diff
@@ -0,0 +1,272 @@
+                        .stream()
+                        .map(BlobMetadata::name)
+                        .collect(Collectors.toList());
+                    try {
+                        blobStoreTransferService.deleteBlobs(
+                            pinnedTimestampsBlobStore.getBlobPathForUpload(remotePinnedTimestamps),
+                            oldFilesToBeDeleted
+                        );
+                    } catch (IOException e) {
+                        logger.error("Exception while deleting stale pinned timestamps", e);
+                    }
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    private PinnedTimestamps readExistingPinnedTimestamps(String blobFilename, RemotePinnedTimestamps remotePinnedTimestamps) {
+        remotePinnedTimestamps.setBlobFileName(blobFilename);
+        remotePinnedTimestamps.setFullBlobName(pinnedTimestampsBlobStore.getBlobPathForUpload(remotePinnedTimestamps));
+        try {
+            return pinnedTimestampsBlobStore.read(remotePinnedTimestamps);
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to read existing pinned timestamps", e);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        asyncUpdatePinnedTimestampTask.close();
+    }
+
+    // Visible for testing
+    public void setPinnedTimestampsSchedulerInterval(TimeValue pinnedTimestampsSchedulerInterval) {
+        this.pinnedTimestampsSchedulerInterval = pinnedTimestampsSchedulerInterval;
+        rescheduleAsyncUpdatePinnedTimestampTask();
+    }
+
+    private void rescheduleAsyncUpdatePinnedTimestampTask() {
+        if (pinnedTimestampsSchedulerInterval != null) {
+            pinnedTimestampsSet = new Tuple<>(-1L, Set.of());
+            asyncUpdatePinnedTimestampTask.close();
+            startAsyncUpdateTask();
+        }
+    }
+
+    public static Tuple<Long, Set<Long>> getPinnedTimestamps() {
+        return pinnedTimestampsSet;
+    }
+
+    /**
+     * Inner class for asynchronously updating the pinned timestamp set.
+     */
+    private final class AsyncUpdatePinnedTimestampTask extends AbstractAsyncTask {
+        private AsyncUpdatePinnedTimestampTask(Logger logger, ThreadPool threadPool, TimeValue interval, boolean autoReschedule) {
+            super(logger, threadPool, interval, autoReschedule);
+            rescheduleIfNecessary();
+        }
+
+        @Override
+        protected boolean mustReschedule() {
+            return true;
+        }
+
+        @Override
+        protected void runInternal() {
+            long triggerTimestamp = System.currentTimeMillis();
+            RemotePinnedTimestamps remotePinnedTimestamps = new RemotePinnedTimestamps(
+                clusterService.state().metadata().clusterUUID(),
+                blobStoreRepository.getCompressor()
+            );
```

Comment:
Right, but the scope of this object is limited so should be collected faster. New object makes it more intuitive rather than using the same object repeatedly. We don't expect timestamp pinning to be frequent for this to create an issue.

Reply:
Let me know what you think. Resolving this conversation, feel free to re-open.